sdk/storage/azblob/internal/exported/structured_message_response_reader.go

```diff
@@ -0,0 +1,191 @@
+			}
+			if errors.Is(err, structuredmsg.ErrUnexpectedEOF) {
+				return 0, fmt.Errorf("unexpected end of structured message: %w", err)
+			}
+			return 0, err
+		}
+		r.headerRead = true
+	}
+
+	// Read data from current segment, or get next segment
+	totalRead := 0
+	for totalRead < len(p) {
+		// If we've exhausted current segment, read next one
+		if r.currentSegmentData == nil || r.currentSegmentIndex >= len(r.currentSegmentData) {
+			segmentData, err := r.reader.ReadSegment()
+			if err != nil {
+				if err == io.EOF {
+					// No more segments, validate trailer
+					_, trailerErr := r.reader.ReadTrailer()
+					if trailerErr != nil {
+						if trailerErr == structuredmsg.ErrCRC64Mismatch {
+							return 0, fmt.Errorf("CRC64 checksum mismatch in structured message trailer: %w", trailerErr)
+						}
+						if trailerErr == structuredmsg.ErrUnexpectedEOF {
+							return 0, fmt.Errorf("unexpected end of structured message trailer: %w", trailerErr)
+						}
+						return totalRead, trailerErr
+					}
+					// Successfully validated all segments and trailer
+					r.closed = true
+					if totalRead == 0 {
+						return 0, io.EOF
+					}
+					return totalRead, nil
+				}
+				if err == structuredmsg.ErrCRC64Mismatch {
+					return 0, fmt.Errorf("CRC64 checksum mismatch in structured message segment: %w", err)
+				}
+				if err == structuredmsg.ErrInvalidSegmentNumber {
+					return 0, fmt.Errorf("invalid segment number in structured message: %w", err)
+				}
+				if err == structuredmsg.ErrUnexpectedEOF {
+					return 0, fmt.Errorf("unexpected end of structured message segment: %w", err)
+				}
+				return totalRead, err
+			}
+			r.currentSegmentData = segmentData
+			r.currentSegmentIndex = 0
+		}
+
+		// Copy data from current segment to output buffer
+		available := len(r.currentSegmentData) - r.currentSegmentIndex
+		needed := len(p) - totalRead
+		copySize := available
+		if copySize > needed {
+			copySize = needed
+		}
+
+		copy(p[totalRead:totalRead+copySize], r.currentSegmentData[r.currentSegmentIndex:r.currentSegmentIndex+copySize])
+		totalRead += copySize
+		r.currentSegmentIndex += copySize
+
+		// If we've consumed the entire segment, mark it as nil for next iteration
+		if r.currentSegmentIndex >= len(r.currentSegmentData) {
+			r.currentSegmentData = nil
+			r.currentSegmentIndex = 0
+		}
+	}
+
+	return totalRead, nil
+}
+
+// Close implements io.Closer. It closes the underlying body and marks the reader as closed.
+// If not all segments have been read, it attempts to read and validate the remaining data.
+func (r *StructuredMessageResponseReader) Close() error {
+	if r.closed {
+		return nil
+	}
+	r.closed = true
+
+	// If we haven't finished reading, try to read remaining segments and validate trailer
+	// This ensures we catch any CRC64 mismatches even if the user doesn't read all data
+	if r.headerRead {
+		// Read any remaining segments
+		for {
+			_, err := r.reader.ReadSegment()
```

Comment:
This is going to create a lot of garbage.  If this is important to do on `Close()` then we need to change the design so that draining the reader doesn't allocate a bunch of `[]byte` that get discarded.